homeassistant/components/automation/__init__.py

```diff
@@ -215,6 +232,109 @@ async def reload_service_handler(service_call):
     return True
 
 
+class AutomationTrace:
+    """Container for automation trace."""
+
+    def __init__(self, unique_id, config, trigger, context, action_trace):
+        """Container for automation trace."""
+        self._action_trace = action_trace
+        self._condition_trace = None
+        self._config = config
+        self._context = context
+        self._error = None
+        self._state = "running"
+        self._timestamp_finish = None
+        self._timestamp_start = dt_util.utcnow()
+        self._trigger = trigger
+        self._unique_id = unique_id
+        self._variables = None
+
+    def set_error(self, ex):
+        """Set error."""
+        self._error = ex
+
+    def set_variables(self, variables):
+        """Set variables."""
+        self._variables = variables
```

Comment:
Variables can change over time and different variable blocks can override them. Are we just storing the final variables?

Reply:
No, variables are also stored in each action (`trace_action`) and each condition (`trace_condition`) too.

https://github.com/home-assistant/core/pull/46755/files#diff-0558c1da8b70e0262af39161c04b536bb9398f7963b41f9338f9e85cee012f42R170

https://github.com/home-assistant/core/pull/46755/files#diff-0f35596b834bf39598024ea5265154ac449f556a73cd7a577053c4a9c61ddaabR218

However, the variables are not included in JSON output yet, I intend to fix that in a follow-up PR:
https://github.com/home-assistant/core/pull/46755/files#diff-6eb1c90205e6041cc1387c3ecba199faac875f9ab7ae9c02e36b2c185cd39259R55-R56

Reply:
Fixed by #47549